docs/DifferentiableProgrammingImplementation.md

```diff
@@ -0,0 +1,1000 @@
+- [Differentiable function types](DifferentiableProgramming.md#differentiable-function-types): subtypes of normal function types, with a different runtime representation and calling convention. Differentiable function types have differentiable parameters and results.
+- [Differential operators](DifferentiableProgramming.md#differential-operators): core differentiation APIs. Differential operators are higher-order functions that take `@differentiable` functions as inputs and return derivative functions or evaluate derivative values.
+- [The differentiation transform](#the-differentiation-transform): a compiler transformation on the Swift Intermediate Language (SIL) that implements automatic differentiation and generates derivative functions. The differentiation transform is explained in this document.
+
+## Terminology
+
+### JVP and VJP functions
+
+A JVP (Jacobian-vector product) function is a [forward-mode](DifferentiableProgramming.md#automatic-differentiation) derivative function. A VJP (vector-Jacobian product) function is a [reverse-mode](DifferentiableProgramming.md#automatic-differentiation) derivative function.
+
+### Typing rules
+
+Consider a function with type `(T0, ...) -> U`, where `T0, ..., U` all conform to the `Differentiable` protocol.
+
+The JVP function type is:
+
+```swift
+//  (T0, ...)      ->  (U,    (T0.TangentVector, ...) -> (U.TangentVector))
+//   ^                  ^      ^~~~~~~~~~~~~~~~~~~~~      ^~~~~~~~~~~~~~~
+//  original args   result      derivative wrt args    derivative wrt result
+```
+
+The JVP function returns a tuple with two elements:
+- The original result of type `U`.
+- A linear approximation function called the “differential”, which takes derivatives with respect to arguments and returns the derivative with respect to the result.
+
+The VJP function type is:
+
+```swift
+//  (T0, ...)      ->  (U,    (U.TangentVector)   -> (T0.TangentVector, ...))
+//   ^                  ^      ^~~~~~~~~~~~~~~        ^~~~~~~~~~~~~~~~~~~~~
+//  original args   result     derivative wrt result    derivative wrt args
+```
+
+The VJP function returns a tuple with two elements:
+- The original result of type `U`.
+- A backpropagation function called the “pullback”, which takes the derivatives with respect to the result and returns the derivative with respect to arguments.
+
+> TODO: Explain typing rules for functions with `inout` parameters.
+
+### Registration
+
+JVP and VJP functions can be registered using the `@derivative` attribute. JVP functions should return a tuple with labels `(value: ..., differential: ...)`. VJP functions should return a tuple with labels `(value: ..., pullback: ...)`.
+
+For usage examples of the `@derivative` attribute, please see [this mailing list response](https://groups.google.com/a/tensorflow.org/d/msg/swift/k62Uwnc-EYc/Cncb7mxtCAAJ), which shows how to use the `@derivative` attribute to register a derivative for `ax^2 + bx + c`.
```

Comment:
I don't think we should refer to mailing list responses to show examples.  Usage of `@derivative` is well-documented in https://github.com/rxwei/swift-evolution/blob/autodiff/proposals/0000-differentiable-programming.md#derivative-attribute and we should link to that instead.

Reply:
Could you provide a suggested change of exactly how you want the new sentence worded?

Reply:
For usage examples of the `@derivative` attribute, please see https://github.com/rxwei/swift-evolution/blob/autodiff/proposals/0000-differentiable-programming.md#derivative-attribute.

Reply:
```suggestion
For usage examples of the `@derivative` attribute, refer to the [`@derivative` attribute section](https://github.com/rxwei/swift-evolution/blob/autodiff/proposals/0000-differentiable-programming.md#derivative-attribute) in the proposal.
```

Reply:
Is this okay now?

Reply:
Yes. Thanks.